### PR TITLE
sync-diff-inspector: fix index fields not taking effect

### LINKED ISSUE #623: sync-diff-inspector: index-fields does not take effect if bucket stats are used.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
```
root@sbtest[v5]> show create table sbtest1 \G
*************************** 1. row ***************************
       Table: sbtest1
Create Table: CREATE TABLE `sbtest1` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `k` int(11) NOT NULL DEFAULT '0',
  `c` char(120) NOT NULL DEFAULT '',
  `pad` char(60) NOT NULL DEFAULT '',
  PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */,
  KEY `k_1` (`k`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin AUTO_INCREMENT=10407284
1 row in set (0.00 sec)
```

```toml
# Diff Configuration.
######################### Global config #########################
check-thread-count = 4
export-fix-sql = true
check-struct-only = false


######################### Datasource config #########################
[data-sources]
[data-sources.upstream]
    host = "upstream-host"
    port = 4000
    user = "root"
    password = ""
    #snapshot = "433361056758235137"

[data-sources.downstream]
    host = "downstream-host"
    port = 5300
    user = "root"
    password = ""
    #snapshot = "433361057203093509"

######################### Task config #########################
[task]
    output-dir = "./output5"
    source-instances = ["upstream"]
    target-instance = "downstream"
    target-check-tables = ["sbtest.sbtest1"]
    target-configs = ["config1"]

######################### Table config #########################
[table-configs.config1]
target-tables = ["sbtest.sbtest1"]
#range = "age > 10 AND age < 20"
index-fields = ["k"]
#ignore-columns = [""]
chunk-size = 50000
#collation = ""

```

2. What did you expect to see?
- The key `k_1` should be used to split chunks.

3. What did you see instead?
- The primary key is used instead.

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
Irrelevant.

6. which tool are you using?
sync-diff-inspector

7. what versionof tool are you using (`pump -V` or `tidb-lightning -V` or `syncer -V`)?
`v5.4.1`.

### PR DESCRIPTION
**Pull request by liuzix**
### What problem does this PR solve?

Issue Number: close #623 

### What is changed and how it works?
- Added `index-fields` index selection for `BucketIterator`

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: sync_diff_inspector/splitter/index_fields.go [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+		log.Panic("matching with empty index")
+	}
+	if len(f.cols) == 0 {
+		log.Panic("unexpected cols with length 0")
+	}
+
+	if len(index.Columns) != len(f.cols) {
+		// We need an exact match.
+		// Lengths not matching eliminates the possibility.
+		return false
```
Comment (lichunzhu) [inline sync_diff_inspector/splitter/index_fields.go:81]: If the given columns can match the first few columns, this index can still work?

Comment (liuzix) [reply sync_diff_inspector/splitter/index_fields.go:81]: The real intention of the user may be to use columns that are part of a composite key. If we allow matching here, it does not do what the user expects.

Comment (lichunzhu) [reply sync_diff_inspector/splitter/index_fields.go:81]: If composite key is `(a,b,c)` and users want to use columns `a,b`, I think we can still match this index? However, `b,a` or `b,a,c` can't be matched.

**Comment by 3pointer** [comment]
I feel like the real problem is we didn't handle `limit` in `bucketSplitter`.  so maybe we should fallback to `randomSplitter` when user set the `range` config. and also we need this PR to handle index-fields properly

**Review by 3pointer — APPROVED** [review]
LGTM

**Review by lichunzhu — APPROVED** [review]
LGTM

**Pull request merged**